### ISSUE TITLE
[MICROGRANT] <Peer-to-peer Imageboard>

### ISSUE THREAD

**Issue by threshold862543** [ipfs/devgrants#253]
### 1. What is your project? (max 100 words)
An imageboard-format forum that uses IPFS for user-uploaded files and OrbitDB to store posts. There are two components to it, one is the "mediator server" which validates new posts and has moderation controls, and the "peer server", which allows a user to manage the data they are interacting with/seeding and render it in the browser. I am planning to further offload functionality from the former to the latter in a multi-phased approach.

Mediator server:
https://gitgud.io/threshold862543/orbitchan
Peer server:
https://gitgud.io/threshold862543/orbitchan-peer

### 2. How will IPFS, Filecoin, or related technology be used for this project? (max 100 words)
IPFS is the main infrastructural component of the project, as it is how all of the user-generated content is stored and distributed among peers.
  
### 3. How will you improve your project with this grant? What steps will you take to meet this objective? (max 200 words)
After speaking with a few developers working on distributed database libraries adjacent to OrbitDB, it seems clear that OrbitDB is no longer being actively developed and that promising alternatives exist that overcome some of OrbitDB's limitations. I want to expand the scope of this project to be able to use these new libraries as well. Two that I want to make this project compatible with are [peerbit](https://github.com/dao-xyz/peerbit) and [opal](https://github.com/cypsela/opal). Further, as this field is rapidly developing, I want to modularize the way the database-interactions work so that new technologies such as those just mentioned can be "slotted in" as they appear, and so users can choose which type of database they want to use during setup when creating their forum, to keep the project abreast of any new innovations. This grant will allow me to work full-time on this refactoring and testing.
 
### 4. Is this project open source?
Everything is open source under the AGPLv3 license.

### 5. Do you agree to share grant reports upon request, including a final grant report at the end of the three month period?
Yes.
  
### 6. Does your proposal comply with our Community Code of Conduct?
Yes.

### 7. Links and submissions
N/A
  
### Additional questions:
* For each team member(s), please list name, email, Github account, and role in the project.
threshold | threshold862543@proton.me | https://github.com/threshold862543 or https://gitgud.io/threshold862543 | sole developer
* How did you learn about our microgrant program?
At IPFS Camp 2022.

**Comment by ErinOCon** [comment]
HI @threshold862543, thank you for your proposal! Do you have a working prototype for this project?

**Comment by threshold862543** [comment]
@ErinOCon Thank you for your response. The current version of the software can be downloaded from the gitgud repository I linked in the proposal. Installation instructions are also provided there. It is in an essentially functional state but I haven't paid for and started hosting an instance on a live server yet because there are still a few things I want to polish up first, and also because I want to revisit the database software in the way I outlined above.

**Comment by ErinOCon** [comment]
HI @threshold862543, thank you for your reply! Does your GitGud repo link to your Github repo in any way? If no, would it be possible to temporarily reference your Github repo in GitGud? Confirmation that these accounts are related will be necessary before approval. 

Please note that this proposal is still under review. Thanks!

**Comment by threshold862543** [comment]
@ErinOCon Thank you for your reply. I added some references and a link to my GitHub account on my GitGud profile, at:

https://gitgud.io/threshold862543

![image](https://user-images.githubusercontent.com/111914967/204413452-f3648eb4-8a89-4cff-99c6-0d414f6ecf2e.png)
  
Is this alright or is there anywhere additional that I should put it? Thanks!

**Comment by ErinOCon** [comment]
Thanks, @threshold862543! Since the project isn't yet live, can you provide a demo video showcasing your building progress?

**Comment by threshold862543** [comment]
Sure. Intentionally it very much looks like and functions like a normal forum, but I'll make a video displaying its functionality. I'm currently traveling so it may take a few days to get the video set up, but I'll upload it as soon as I can. I hope that's alright.

**Comment by threshold862543** [comment]
Here are two clips demonstrating the use of the software.

This first clip demonstrates the creation of a new thread via a traditional POST request. The mediator server processes the poster's request, adds the user's file to IPFS, and then adds the post data, including things like comment text and IPFS file CID, as a new entry in the OrbitDB database:



https://user-images.githubusercontent.com/111914967/205293379-73503db1-c60a-4e0e-9529-d16dc08ee43b.mp4


This second clip demonstrates the browsing of newly made thread via orbitchan peer, that is to say, retrieving the post and file data via IPFS and OrbitDB.  This is then rendered and page-served locally via the orbitchan peer server. The user then replies to the thread with their own image, using a POST request, and then the new post data and image is retrieved and rendered again via IPFS:

https://user-images.githubusercontent.com/111914967/205293450-d5414835-60e4-46bd-ab9f-4c5d049faec8.mp4

So, in summary, this demonstrates the software being used to share and replicate files and post data in a peer-to-peer way over IPFS and OrbitDB.

**Comment by threshold862543** [comment]
Thank you very much. That's great news.
